File: timetable.py

import sys
import os
import datetime




def main():
    filenm = sys.argv[1]
    # current time currently hard coded for testing purposes.
    # this will come as sys.argval[2]
    # also when current time is recieve, reset the seconds to 00 so it will also return the exact time match. 
    # e.g 10.03.55 will not match 10.03 in text file. the incomming time second should be rounded.
    currentTime = datetime.time(hour=10, minute=3, second=00)

    
    print(f"Current time is: {currentTime}")

    # createTime= time.ctime (os.path.getctime(filenm))
    modT= os.path.getmtime(filenm)

    #load the file name
    timetable=loadfile(filenm)
    

    # update file
    if (modT != os.path.getmtime(filenm)):
        print("Timetable Updated\n")
        timetable = update(filenm)

    #store the earliest timetable in a list
    paths = earliest(currentTime, timetable)

    # printing for testing purposes. This will show what it returns
    for entry in paths:
        print(entry)

    print("\n") 
    for entry in timetable:
        print(entry)
    return paths, timetable



def loadfile(filename):
    #list of dictionary timetable to return storing depart time, route name, depart from, arrival time, arrival station
    timetable=[]
    with open (filename) as fread:

        # header of timetables, can remove if not needed. if removed modify earliest function to include the very first entry
        for line in fread:
            if not line.startswith("#"):
                stationInfo= line.strip().split(',')
                timetable.append({
                    'StationName': stationInfo[0],
                    'Longitude': stationInfo[1],
                    'Latitude': stationInfo[2],
                })
                break

        # actual timetable data
        for line in fread:
            if not line.startswith("#") and line.strip():
                datarow = line.strip().split(',')
                timetable.append({
                    'departTime':datarow[0],
                    'routeName': datarow[1],
                    'departFrom': datarow[2],
                    'arriveTime': datarow[3],
                    'arriveAt': datarow[4]
                })
        return timetable

#check happens in main. if file has been modified. reload the file by rereading the function
def update(filename):    
    return loadfile(filename)

def earliest(curTime, timetable):
    
    unique_destinations = {}

    #The for loop will do the following
    #filter the times that are not the past
    #if the destination is not in the unique_destinations entry add it
    #otherwise check the uniq_destination entry and update if required (keeping earliest time)
    for entry in timetable[1:]:
        destination = entry['arriveAt']
        depart_time = entry['departTime']
        depart_time_obj = datetime.datetime.strptime(depart_time, '%H:%M').time()   #convert text time into datetime time
        if depart_time_obj >= curTime:     
            if destination not in unique_destinations:
                unique_destinations[destination] = entry
            else:
                if depart_time < unique_destinations[destination]['departTime']:
                    unique_destinations[destination] = entry
                                                              

    return list(unique_destinations.values()) #convert back to list



if __name__ == "__main__":
    main()

File: timetable.c
#define _POSIX_C_SOURCE 200809L
#define _XOPEN_SOURCE       /* See feature_test_macros(7) */


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define MAX_TIMETABLE_ENTRIES 100

struct timetable_entry {
    char departTime[6]; // Assuming time format HH:MM
    char routeName[20];
    char departFrom[20];
    char arriveTime[6]; // Assuming time format HH:MM
    char arriveAt[20];
};

struct timetable_entry timetable[MAX_TIMETABLE_ENTRIES];

void loadfile(const char *filename) {
    FILE *file = fopen(filename, "r");
    if (file == NULL) {
        perror("Error opening file");
        exit(EXIT_FAILURE);
    }

    char line[100];
    int index = 0;

    // Read header
    if (fgets(line, sizeof(line), file) != NULL) {
        // Assuming format: StationName,Longitude,Latitude
        char *token = strtok(line, ",");
        strcpy(timetable[index].departTime, token);
        token = strtok(NULL, ",");
        strcpy(timetable[index].routeName, token);
        token = strtok(NULL, ",");
        strcpy(timetable[index].departFrom, token);
        token = strtok(NULL, ",");
        strcpy(timetable[index].arriveTime, token);
        token = strtok(NULL, ",");
        strcpy(timetable[index].arriveAt, token);
        index++;
    }

    // Read timetable data
    while (fgets(line, sizeof(line), file) != NULL && index < MAX_TIMETABLE_ENTRIES) {
        // Assuming format: departTime,routeName,departFrom,arriveTime,arriveAt
        char *token = strtok(line, ",");
        strcpy(timetable[index].departTime, token);
        token = strtok(NULL, ",");
        strcpy(timetable[index].routeName, token);
        token = strtok(NULL, ",");
        strcpy(timetable[index].departFrom, token);
        token = strtok(NULL, ",");
        strcpy(timetable[index].arriveTime, token);
        token = strtok(NULL, ",");
        strcpy(timetable[index].arriveAt, token);
        index++;
    }

    fclose(file);
}

void earliest(const char *curTime) {
    struct tm tm_curTime;
    strptime(curTime, "%H:%M", &tm_curTime);

    printf("Current time is: %s\n", curTime);

    for (int i = 1; i < MAX_TIMETABLE_ENTRIES; i++) {
        struct tm tm_departTime;
        strptime(timetable[i].departTime, "%H:%M", &tm_departTime);
        if (difftime(mktime(&tm_departTime), mktime(&tm_curTime)) >= 0) {
            printf("Destination: %s\n", timetable[i].arriveAt);
            printf("Departure time: %s\n", timetable[i].departTime);
            break; // Assuming you only want to print the earliest destination
        }
    }
}

int main(int argc, char *argv[]) {
    if (argc != 3) {
        fprintf(stderr, "Usage: %s <filename> <current time (HH:MM)>\n", argv[0]);
        exit(EXIT_FAILURE);
    }

    const char *filename = argv[1];
    const char *curTime = argv[2];

    loadfile(filename);
    earliest(curTime);

    return 0;
}
